components/button/button.tsx

```diff
@@ -50,9 +48,16 @@ export interface BaseButtonProps {
   ghost?: boolean;
 }
 
-export type AnchorButtonProps = BaseButtonProps & React.AnchorHTMLAttributes<HTMLAnchorElement>;
+export type AnchorButtonProps = {
+  href: string;
+  target?: string;
+  onClick?: React.MouseEventHandler<HTMLAnchorElement>;
+} & BaseButtonProps & React.AnchorHTMLAttributes<HTMLAnchorElement>;
 
-export type NativeButtonProps = BaseButtonProps & React.ButtonHTMLAttributes<HTMLButtonElement>;
+export type NativeButtonProps = {
+  htmlType?: string;
+  onClick?: React.MouseEventHandler<HTMLButtonElement>;
+} & BaseButtonProps & React.ButtonHTMLAttributes<HTMLButtonElement>;
 
 export type ButtonProps = AnchorButtonProps | NativeButtonProps;
```

Comment:
@yesmeck What's the roadmap to introducing Conditional Type into the code base? We can improve the Omit type in `components/_util/type.ts`, and for better intelligence benefits, we can use

(thanks to https://github.com/Microsoft/TypeScript/issues/12936#issuecomment-393202175)
```typescript
type AllKeys<U> = U extends any ? keyof U : never;
type ExclusifyUnion<U> = [U] extends [infer V] ?
 V extends any ?
 (V & {[P in Exclude<AllKeys<U>, keyof V>]?: never})
 : never : never;

type ButtonProps = ExclusifyUnion<AnchorButtonProps | NativeButtonProps>;
``` 
To narrow untion props in calling them.

Once we use Button as
```tsx
<Button href="/path/to/open" />
```
It doesn't receive 'htmlType' as prop anymore.

Reply:
We can use conditional type.

Reply:
@yesmeck Start it from replace the old `Omit` advanced type implementation with `Exclude` way? I can take a PR for it.

Reply:
https://github.com/ant-design/ant-design/blob/master/components/_util/type.ts

Reply:
We're already using `Exclude`.

Reply:
Cool...